examples/dl-capsule/sqrl.py

```diff
@@ -1,6 +1,120 @@
 import torch
 import ksc.torch_frontend as knossos
 
+import ksc.compile
+from ksc.torch_frontend import cpp_string_to_autograd_function
+
+embedded_cflags = ksc.compile.default_cflags
+
+
+embedded_cflags_opts = ksc.compile.CFlags.GCCOnly(
+    ["-march=native", "-funroll-loops", "-ffast-math", "-mprefer-vector-width=512",]
+)
+
+
+cpp_inlined_map = """
+        #include "knossos.h"
+
+        namespace ks{
+        tensor<2, ks::Float> sqrl(ks::allocator * $alloc, tensor<2, ks::Float> t) {
+            auto tdata = t.data();
+            ks::Float sum = 0.0;
+            auto ne = t.num_elements();
+
+            for (int i = 0; i != ne; ++i) {
+                sum += tdata[i];
+            }
+
+            ks::Float outsum = 0.0;
+
+            if (sum < 0.0) {
+                for (int i = 0; i != ne; ++i) {
+                    ks::Float r = -0.125 * tdata[i];
+                    outsum += sin(r) * r;
+                }
+            } else {
+                for (int i = 0; i != ne; ++i) {
+                    ks::Float r = 0.5 * tdata[i] * tdata[i];
+                    outsum += sin(r) * r;
+                }
+            }
+
+            ks::Float outmean = outsum / ne;
+
+            auto outtensor = tensor<2, ks::Float>::create($alloc, make_Tuple(1, 1));
+            outtensor.data()[0] = outmean;
+            return outtensor;
```

Comment:
The ksc-generated version returns a `ks::Float`, so I think this should probably do the same. What the return type of `entry` should be is another matter ... the generated entry point returns a `ks::Float` still, but arguably we should convert it to a 0-dimensional Tensor to match PyTorch. I don't think a 1x1 Tensor is right though, and I'm slightly surprised that it works!

Reply:
I did get some segfaults on larger sizes, so perhaps that is why! Can you elaborate concretely on what I should do?

Reply:
The entry points generated by `cgen.py` currently look like this:

```
ks::Float entry(torch::Tensor arg0) {
    auto ks_arg0 = convert_argument<ks::tensor<2, ks::Float>>(arg0);

    auto ks_ret = ks::sqrl$aT2f(&g_alloc , ks_arg0);

    return convert_return_value<ks::Float>(ks_ret);
}

torch::Tensor entry_vjp(torch::Tensor arg0, ks::Float arg1) {
    auto ks_arg0 = convert_argument<ks::tensor<2, ks::Float>>(arg0);
    auto ks_arg1 = convert_argument<ks::Float>(arg1);

    auto ks_ret = ks::sufrev$sqrl$aT2f(&g_alloc , ks_arg0, ks_arg1);

    return convert_return_value<torch::Tensor>(ks_ret);
}
```

I'd suggest using similar code for `embedded_cpp_entry_points`, changing `sqrl` to return a `ks::Float` (ie. just return `outmean` directly, rather than wrapping it in a tensor). Similarly the second argument to `sufrev$sqrl` is a `ks::Float` in the generated code.

Reply:
I'm doubtful that this would explain the segfaults though, unfortunately!